pkg/e2e/autoscaler/autoscaler.go

```diff
@@ -173,165 +182,144 @@ func machineAutoscalerResource(targetMachineSet *mapiv1beta1.MachineSet, minRepl
 
-		initialNumberOfReplicas0 := pointer.Int32PtrDerefOr(targetMachineSet0.Spec.Replicas, e2e.DefaultMachineSetReplicas)
-		initialNumberOfReplicas1 := pointer.Int32PtrDerefOr(targetMachineSet1.Spec.Replicas, e2e.DefaultMachineSetReplicas)
-		glog.Infof("initialNumberOfReplicas0 %d, initialNumberOfReplicas1 %d", initialNumberOfReplicas0, initialNumberOfReplicas1)
-
-		g.By("Creating workload")
-		err = client.Create(context.TODO(), newWorkLoad())
+		g.By("Getting nodes")
+		nodes, err := e2e.GetNodes(client)
 		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(len(nodes)).To(o.BeNumerically(">=", 1))
+
+		g.By(fmt.Sprintf("Creating %v machineautoscalers", len(machineSets)))
+		var clusterExpansionSize int
+		for i := range machineSets {
+			o.Expect(pointer.Int32PtrDerefOr(machineSets[i].Spec.Replicas, 0)).To(o.BeNumerically("==", 1))
+			var min, max int32 = 1, 2
+			glog.Infof("Create MachineAutoscaler backed by MachineSet %s/%s - min:%v, max:%v", machineSets[i].Namespace, machineSets[i].Name, min, max)
+			asr := machineAutoscalerResource(&machineSets[i], min, max)
+			o.Expect(client.Create(context.TODO(), asr)).Should(o.Succeed())
+			cleanupObjects = append(cleanupObjects, runtime.Object(asr))
+			clusterExpansionSize += int(max - min)
+		}
+		o.Expect(clusterExpansionSize).To(o.BeNumerically(">", 1))
+
+		// We want to scale out to max-cluster-size-1. We
+		// choose max-1 because we want to test that
+		// maxNodesTotal is respected by the
+		// cluster-autoscaler. If maxNodesTotal ==
+		// max-cluster-size then no MaxNodesTotalReached
+		// event will be generated.
+		maxNodesTotal := len(nodes) + clusterExpansionSize - 1
+
+		eventWatcher := newEventWatcher(clientset)
+		o.Expect(eventWatcher.run()).Should(o.BeTrue())
+		defer eventWatcher.stop()
+
+		// Log cluster-autoscaler events
+		eventWatcher.onEvent(matchAnyEvent, func(e *corev1.Event) {
+			if e.Source.Component == clusterAutoscalerComponent {
+				glog.Infof("%s: %s", e.InvolvedObject.Name, e.Message)
+			}
+		}).enable()
+
+		g.By(fmt.Sprintf("Creating ClusterAutoscaler configured with maxNodesTotal:%v", maxNodesTotal))
+		clusterAutoscaler := clusterAutoscalerResource(maxNodesTotal)
+		o.Expect(client.Create(context.TODO(), clusterAutoscaler)).Should(o.Succeed())
+		cleanupObjects = append(cleanupObjects, runtime.Object(clusterAutoscaler))
+
+		g.By("Creating scale-out workload")
+		scaleUpCounter := newScaleUpCounter(eventWatcher, uint32(clusterExpansionSize-1))
+		workload := newWorkLoad()
+		o.Expect(client.Create(context.TODO(), workload)).Should(o.Succeed())
+		cleanupObjects = append(cleanupObjects, runtime.Object(workload))
+		o.Eventually(func() uint32 {
+			v := scaleUpCounter.get()
+			glog.Infof("Waiting for %s to generate %v more %q events", clusterAutoscalerComponent, v, clusterAutoscalerScaledUpGroup)
+			return v
+		}, e2e.WaitLong, 3*time.Second).Should(o.BeZero())
+
+		maxNodesTotalReachedCounter := newMaxNodesTotalReachedCounter(eventWatcher, 1)
```

Comment:
What if you observe `Max total nodes in cluster reached` event before the last `Scale-up: setting group`? Then, this will timeout.

Reply:
Then that would imply a bug in the cluster-autoscaler.

Reply:
The CA will keep sending `MaxNodesTotalReached` events - it is not just a one-shot event. Everytime `--scan-interval` it will try to scale up but this condition will be met and so send another event.